docs/writing_a_builder.md

```diff
@@ -31,6 +31,48 @@ will be passed as a primary input, and each build step may produce two assets.
 For the primary input `some_file.txt` the allowed outputs are
 `some_file.txt.foo` and `some_file.txt.bar`.
 
+Keys in `buildExtensions` match a suffix in the path of potential inputs. That
+is, a builder will run when an input ends with its input extension.
+Valid outputs are formed by replacing the matched suffix with values in that 
+map. For instance, `{'.dart': ['.g.dart']}` matches all files ending with
+`.dart` and allows the builder to write a file  with the same name but with a
+`.g.dart` extension instead.
+
+Builders can declare more complex inputs and outputs by using a capture group
```

Comment:
I think it would be good to add a header for this section:

```suggestion
### Capture Groups

Builders can declare more complex inputs and outputs by using a capture group
```